lib/adal.js

```diff
@@ -316,13 +320,37 @@ AuthenticationContext.prototype._renewIdToken = function (callback) {
     this.verbose('Navigate to:' + urlNavigate);
     this._saveItem(this.CONSTANTS.STORAGE.LOGIN_REQUEST, '');
     frameHandle.src = 'about:blank';
-    this._loadFrame(urlNavigate, 'adalIdTokenFrame');
+    this._loadFrameTimeout(urlNavigate, 'adalIdTokenFrame', this.config.clientId, this.CONSTANTS.LOADFRAME_TIMEOUT);
 };
 
 AuthenticationContext.prototype._urlContainsQueryStringParameter = function (name, url) {
     // regex to detect pattern of a ? or & followed by the name parameter and an equals character
     var regex = new RegExp("[\\?&]" + name + "=");
     return regex.test(url);
+};
+
+/* Calling _loadFrame but with a timeout to signal failure in loadframeStatus. Frames are left
+ * hanging when network errors occur.
+ * https://github.com/AzureAD/azure-activedirectory-library-for-js/issues/250.
+ * */
+AuthenticationContext.prototype._loadFrameTimeout = function (urlNavigation, frameName, resource, timeout) {
+    //set iframe session to pending
+    this._logstatus('Set loading state to pending for: ' + resource);
+    window.loadframeStatus[resource] = this.CONSTANTS.LOADFRAME_PENDING;
+
+    this._loadFrame(urlNavigation, frameName);
+
+    var timeoutDuration = timeout/1000;
+    var resourceSaved = resource;
+    var self = this;
+
+    setTimeout(function () {
+        if (window.loadframeStatus[resource] === self.CONSTANTS.LOADFRAME_PENDING) {
+            //fail the iframe session if it's in pending state
+            self._logstatus('Loading frame has timed out after: ' + timeoutDuration + ' seconds for resource ' + resourceSaved);
+            window.loadframeStatus[resource] = self.CONSTANTS.LOADFRAME_EXPIRED;
```

Comment:
we also need to invoke the callbacks to let them know that there was an error/timeout